includes/blueprints/import.php

```
@@ -0,0 +1,471 @@
+			$term_ids_old_new[ $term['term_id'] ] = $inserted_term['term_id'];
+		}
+	}
+
+	return $term_ids_old_new;
+}
+
+
+/**
+ * Sets terms on posts.
+ *
+ * @param array $post_terms Post term data.
+ * @param array $post_ids_old_new A map of original post IDs from the manifest
+ *                                and their new ID when imported.
+ * @param array $term_ids_old_new A map of original term IDs from the manifest
+ *                                and their new ID when imported.
+ * @return true|WP_Error True on success, WP_Error if setting any term failed.
+ */
+function tag_posts( array $post_terms, array $post_ids_old_new, array $term_ids_old_new ) {
+	foreach ( $post_terms as $post_id => $terms ) {
+		foreach ( $terms as $term ) {
+			$new_post_id = $post_ids_old_new[ $post_id ] ?? $post_id;
+			$new_term_id = $term_ids_old_new[ $term['term_id'] ] ?? $term['term_id'];
+			$result      = wp_set_post_terms( $new_post_id, [ (int) $new_term_id ], $term['taxonomy'], true );
+
+			if ( is_wp_error( $result ) ) {
+				return $result;
+			}
+		}
+	}
+
+	return true;
+}
+
+/**
+ * Imports media files.
+ *
+ * @param array  $media Paths to media files keyed by original media ID, relative
+ *                      to the folder they were unzipped in.
+ * @param string $blueprint_folder Path to the blueprint folder.
+ * @return array|WP_Error Map of old and new media IDs on success.
+ */
+function import_media( array $media, string $blueprint_folder ) {
+	/**
+	 * Stores media IDs that changed during import.
+	 */
+	$media_ids_old_new = [];
+
+	foreach ( $media as $original_media_id => $relative_file_path ) {
+		$full_file_path = $blueprint_folder . '/' . $relative_file_path;
+
+		if ( ! is_readable( $full_file_path ) ) {
+			return new WP_Error(
+				'acm_media_error',
+				sprintf(
+					// translators: the full path to the media file.
+					__( 'Could not read media file at %s.', 'atlas-content-modeler' ),
+					$full_file_path
+				)
+			);
```

I think media import errors probably should result in immediate import failure, but open to opinions on this one too. (We could warn about the missing image and move on, but it's likely to cause broken images once the import completes.)

---

This is a little tricky because it actually can break the content, but at the same time I would like the import to continue. Would it be possible down the road to allow for a `flag` to either ignore the errors or exit?

---

Oh, nice idea to add a flag at some point to make errors soft/hard fail.

Thanks for the feedback! I'll adjust the failure behavior of the others above at least.

---

Leaving media import the same for now. We can add a flag in the future if there's demand for this.